KEYCLOAK-15015 Extend KeyWrapper to add whole certificate chain in x5c parameter

Xiangjiaox opened the pull request.

Commit:
KEYCLOAK-15015 Extend KeyWrapper to add whole certificate chain in x5c parameter and validate certificate chain against truststore.

GisleGrimen commented:
Looks good. Just what we need

pedroigor reviewed (commented):
@Xiangjiaox The idea makes sense.

But, how are you actually setting the chain? Don't you need to consider loading the chain from a truststore, for instance?

None of our RSA key providers are setting it.

Xiangjiaox commented:
@pedroigor My thinking was that one would choose to load the chain when implementing the loadKey method from AbstractRsaKeyProvider. In retrospect i realize it might not be that good to implement a solution that is not available by default, unless you create your own keyprovider.

Should i add logic for retrieving the chain in JavaKeystoreKeyProvider and ImportedRsaKeyProvider?

pedroigor commented:
@Xiangjiaox I was afraid to be missing something :) Thanks.

I think so. However, we should consider the specs and what they say about `x5c`. For instance, the first in the list should be the certificate itself, followed in order by others.

Xiangjiaox commented:
Sorry for the delay, it's been a busy couple of weeks.

I've updated the PR with a suggested implementation.

Having some issues with the commons module dependency since i made changes to PemUtils. Didn't quite see from the docs how to rebuild and update that dependency. Any tips?

Right now I've implemented it so that it expects that the certificate you load will also be present as the first element in the certificate chain. That might be a bit redundant (in the importedRsaKeyProvider), but I'm open to changing it if you think it's better to only load the intermediate certs in the chain and prepending the loaded certificate.
From what I've read on the Keystore api, it will always return the user certificate and the chain (if present) so it is at least consistent between JavaKeystoreKeyProvider and ImportedRsaKeyProvider this way.

Do you think we should validate the certificate with the loaded certificate chain?

pedroigor commented on services/src/main/java/org/keycloak/keys/JavaKeystoreKeyProvider.java:
```diff
@@ -62,7 +59,12 @@ protected KeyWrapper loadKey(RealmModel realm, ComponentModel model) {
+            final List<X509Certificate> certificateChain = Optional.ofNullable(keyStore.getCertificateChain(model.get(JavaKeystoreKeyProviderFactory.KEY_ALIAS_KEY)))
+                    .map(certificates -> (X509Certificate[]) certificates)
+                    .map(Arrays::asList)
+                    .orElseGet(ArrayList::new);
```
This should be better `orElseGet(Collections::emptyList)`.

pedroigor commented on services/src/main/java/org/keycloak/keys/AbstractRsaKeyProvider.java:
```diff
@@ -66,6 +62,10 @@ public AbstractRsaKeyProvider(RealmModel realm, ComponentModel model) {
     }
 
     protected KeyWrapper createKeyWrapper(KeyPair keyPair, X509Certificate certificate) {
+        return createKeyWrapper(keyPair, certificate, new ArrayList<>());
```
Use `Collections.emptyList()`.

pedroigor reviewed (changes requested):
There is an error when trying to persist a new key provider due to the size of the `COMPONENT_CONFIG.VALUE` column.

We should increase this value to be able to support certificate chains using the `ImportedRsaKeyProvider` provider.

Ideally, we should also add some test to cover this. Please, take a look at `ImportedRsaKeyProviderTest`.

pedroigor commented:
> Having some issues with the commons module dependency since i made changes to PemUtils. Didn't quite see from the docs how to rebuild and update that dependency. Any tips?

Sorry. Not sure what you mean.

> Right now I've implemented it so that it expects that the certificate you load will also be present as the first element in the certificate chain. That might be a bit redundant (in the importedRsaKeyProvider), but I'm open to changing it if you think it's better to only load the intermediate certs in the chain and prepending the loaded certificate.

I think it is OK to expect the chain to use the correct order. It is kind of expected anyways whenever you are creating a PEM file for your chain. Perhaps, we might prepend the cert if not there in the chain too. I would also keep the check you did and probably add more info to the "tip" of the new field about the importance of ordering and that the server cert should be there (and if not, it will be added).

> From what I've read on the Keystore api, it will always return the user certificate and the chain (if present) so it is at least consistent between JavaKeystoreKeyProvider and ImportedRsaKeyProvider this way.

+1

> 
> Do you think we should validate the certificate with the loaded certificate chain?

That is a good point. I think it makes sense, yeah. So that we won't accept any invalid cert.

pedroigor commented:
@Xiangjiaox Looks like we have some test failures. Could you rebase your branch? And make sure the failing `JavaKeystoreKeyProviderTest` test is working for you?

Xiangjiaox commented:
Sounds good. I'm going to set of time to implement the suggestions and fix the tests the coming weekend

pedroigor commented:
@Xiangjiaox Please look at https://github.com/pedroigor/keycloak/tree/KEYCLOAK-15015.

Xiangjiaox commented:
I've added certificate chain validation in the validateConfiguration method of ImportedRsaKeyProviderFactory. Is that OK or should we also/or validate it in the constructor of AbstractRsaKeyProvider?
I'm also curious if you know whether the dependency on the Truststore provider will cause issues with regards to instantiation? I.e. is is guaranteed that the truststore will be instantiated before it reaches the key provider factories.

I extended COMPONENT_CONFIG.VALUE to 10k which will allow max 5 pem encoded certs. Is that to low? 

With regards to the tests I'm having some issues running them locally. Currently looking into it

Xiangjiaox commented on services/src/main/java/org/keycloak/keys/ImportedRsaKeyProviderFactory.java:
```diff
@@ -84,6 +88,35 @@ public void validateConfiguration(KeycloakSession session, RealmModel realm, Com
+                try {
+                    final KeyStore truststore = session.getProvider(TruststoreProvider.class).getTruststore();
+                    PKIXParameters params = new PKIXParameters(truststore);
+                    params.setRevocationEnabled(false);
```
I'm assuming here that it is valid to have a root certificate in the certificate chain.
Related: https://stackoverflow.com/a/38128883/11347438

pedroigor commented:
> I've added certificate chain validation in the validateConfiguration method of ImportedRsaKeyProviderFactory. Is that OK or should we also/or validate it in the constructor of AbstractRsaKeyProvider?

Would be better to have the same validation for both providers: keystore and imported.

> I'm also curious if you know whether the dependency on the Truststore provider will cause issues with regards to instantiation? I.e. is is guaranteed that the truststore will be instantiated before it reaches the key provider factories.

Yes, it is. The validation happens at runtime rather than startup (when the trustore is initialized).

> 
> I extended COMPONENT_CONFIG.VALUE to 10k which will allow max 5 pem encoded certs. Is that to low?

To be honest. I would not change that all. But only support a certificate chain when using a truststore. At least for now. I would like to have these changes with only the bare minimum without touch storage.

WDYT? The size of this column is increased from time to time. We should probably try to come up with something better.

> 
> With regards to the tests I'm having some issues running them locally. Currently looking into it

Check my last comment and the branch I linked.

Also, we need to make sure that tests are actually using a chain. Instead of only the leaf cert.

Xiangjiaox commented:
> To be honest. I would not change that all. But only support a certificate chain when using a truststore. At least for now. I would like to have these changes with only the bare minimum without touch storage.
WDYT? The size of this column is increased from time to time. We should probably try to come up with something better.

I agree. It seems like more work for a solution that probably will be used by people who either use a truststore or will implement their own provider. What do you suggest, should we remove the option to load a chain for the Imported provider and just add the chain validation if a chain is present i.e for Keystore providers?

pedroigor commented:
@Xiangjiaox Yes. I think it is better for now. We can always change that in the future.

pedroigor commented:
Also, please consider adding a cert chain to the keystore used during tests. Otherwise, we are not really covering it.

Xiangjiaox commented:
I've implemented certificate chain validation for the JavaKeystoreProvier and added an additional certificate signed by the selfsigned certificate already present.

I've read in the KC documentation that if a truststore isn't set then it will fall back to the standard java truststore, but is that also instantiated as a TrustStoreProvider? In other words could `session.getProvider(TruststoreProvider.class).getTruststore();` ever be null?
I've tested the certificate loading separately as i can't seem to get the testsuite running on my computer. I also haven't found a good way to test the actual certificate validation, I've only done negative tests. So I am open to removing the validation feature if you think that's best.

pedroigor committed:
[KEYCLOAK-15015] - Publishing the x5c for JWK

pedroigor commented:
@Xiangjiaox Please, take a look at https://github.com/pedroigor/keycloak/tree/KEYCLOAK-15015. If you include that I think we should be good to go.

pedroigor commented:
@Xiangjiaox We should also have some mention in docs saying that chain is expected to be imported into the same alias used to load the key pair.

See https://www.keycloak.org/docs/latest/server_admin/#loading-keys-from-a-java-keystore.

Xiangjiaox commented:
I've incorporated the changes you made, thanks for the help! Let me know if you need me to merge the commits, I'm new to the whole git merge rebase workflow.

Created a PR with a small paragraph describing the chain loading: https://github.com/keycloak/keycloak-documentation/pull/1144.
I see that the build is failing, but it looks like that is the case for most of the PRs. Assuming that one line change wont break any tests so maybe something is broken there? I guess we'll continue that conversation there :)

pedroigor commented:
@Xiangjiaox  Thank you.
